### PR TITLE
Expose WebHostBuilderContext in UseKestrel

### PR DESCRIPTION
**Pull request by Tratcher**
#1334 @halter73 I've poached this because it will make the configuration work easier.

### CODE REVIEW

File: src/Kestrel/WebHostBuilderKestrelExtensions.cs [hunk 1]
```diff
@@ -57,5 +57,31 @@ public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder, Actio
...
+        /// <param name="hostBuilder">
+        /// The Microsoft.AspNetCore.Hosting.IWebHostBuilder to configure.
+        /// </param>
+        /// <param name="configureOptions">A callback to configure Kestrel options.</param>
+        /// <returns>
+        /// The Microsoft.AspNetCore.Hosting.IWebHostBuilder.
+        /// </returns>
+        public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder, Action<WebHostBuilderContext, KestrelServerOptions> configureOptions)
+        {
+            if (configureOptions == null)
```
Comment (JunTaoLuo) [inline src/Kestrel/WebHostBuilderKestrelExtensions.cs]: Should this null check be added to the extension above as well?

Comment (Tratcher) [reply src/Kestrel/WebHostBuilderKestrelExtensions.cs]: services.Configure does that check.

File: samples/SampleApp/Startup.cs [hunk 2]
```diff
@@ -57,8 +57,8 @@ public static Task Main(string[] args)
                 {
                     factory.AddConsole();
                 })
-                .UseKestrel(options =>
                 {
+                .UseKestrel((context, options) =>
```
Comment (JunTaoLuo) [inline samples/SampleApp/Startup.cs]: This doesn't look right

**Comment by davidfowl** [comment]
I wonder if this would be better to put on the KestrelServerOptions object itself.

**Comment by Tratcher** [comment]
```
                .UseKestrel(options =>
                  {
                    var basePort = options.HostContext.Configuration.GetValue<int?>("BASE_PORT") ?? 5000;
                    if (options.HostContext.HostingEnvironment.IsDevelopment()) { ... }
```

*shrug*

It may mess up `services.Configure<KestrelServerOptions>(options => ...)`, or you'd have to capture it like you capture the services today.

**Comment by davidfowl** [comment]
> It may mess up services.Configure<KestrelServerOptions>(options => ...), or you'd have to capture it like you capture the services today.

Is the WebHostBuilderContext in the container?

**Comment by Tratcher** [comment]
No

**Comment by davidfowl** [comment]
Then this is fine

**Pull request merged**